examples/vnet/doca_gw.c

```diff
@@ -159,21 +199,47 @@ int doca_gw_port_stop(struct doca_gw_port *port)
     return 0;
 }
 
-
 struct doca_gw_pipeline *doca_gw_create_pipe(struct doca_gw_pipeline_cfg *cfg, struct doca_gw_error *err)
 {
     static uint32_t pipe_id = 1;
-    struct doca_gw_pipeline *pl = malloc(sizeof(struct doca_gw_pipeline));
-    memset(pl,0,sizeof(struct doca_gw_pipeline));
+    struct doca_gw_pipeline *pl;
 
-    if (cfg != NULL && pl != NULL) {
-        // allocate what is needed
-        pl->id = pipe_id++;
-    }
+	pl = malloc(sizeof(struct doca_gw_pipeline));
+    if (cfg == NULL || pl == NULL)
+		return NULL;
+    memset(pl,0,sizeof(struct doca_gw_pipeline));
+	LIST_INIT(&pl->entry_list);
+	pl->id = pipe_id++;
     pl->handler = doca_gw_dpdk_create_pipe(cfg, err);
     return pl;
 }
 
+static int doca_gw_free_pipe(uint16_t portid, struct doca_gw_pipeline *pipe)
+{
+	struct doca_gw_pipelne_entry *entry = LIST_FIRST(&pipe->entry_list);
+
+	while (entry) {
+		LIST_REMOVE(entry, next);
+
+		doca_gw_dpdk_pipe_free_entry(portid, entry);
+		free(entry);
+		entry = LIST_FIRST(&pipe->entry_list);
+	}
+
+	free(pipe);
+	return 0;
+}
+
+void doca_gw_destroy(uint16_t port_id)
```

Comment:
you should call dpdk layer, in dpdk layer
for_each(entry, pipe,....)
   release_entry(....)

release(pipe)

Reply:
same, another commit.